Userland/DevTools/HackStudio/ProjectTemplate.cpp

```diff
@@ -0,0 +1,276 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "ProjectTemplate.h"
+#include <AK/LexicalPath.h>
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/ConfigFile.h>
+#include <LibCore/DirIterator.h>
+#include <LibCore/File.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <spawn.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+// FIXME: shameless copy+paste from Userland/cp. We should have system-wide file management functions.
+// Issue #5209
+bool copy_file_or_directory(String, String, bool, bool);
+bool copy_file(String, String, const struct stat&, int);
+bool copy_directory(String, String, bool);
+
+namespace HackStudio {
+
+ProjectTemplate::ProjectTemplate(const String& id, const String& name, const String& description, const GUI::Icon& icon, int priority)
+    : m_id(id)
+    , m_name(name)
+    , m_description(description)
+    , m_icon(icon)
+    , m_priority(priority)
+{
+}
+
+RefPtr<ProjectTemplate> ProjectTemplate::load_from_manifest(const String& manifest_path)
+{
+    auto config = Core::ConfigFile::open(manifest_path);
+
+    if (!config->has_group("HSTemplate")
+        || !config->has_key("HSTemplate", "Name")
+        || !config->has_key("HSTemplate", "Description"))
+        return {};
+
+    auto id = LexicalPath(manifest_path).title();
+    auto name = config->read_entry("HSTemplate", "Name");
+    auto description = config->read_entry("HSTemplate", "Description");
+    int priority = config->read_num_entry("HSTemplate", "Priority", 0);
+
+    // Attempt to read in the template icons
+    // Fallback to a generic executable icon if one isn't found
+    auto icon = GUI::Icon::default_icon("filetype-executable");
+
+    auto bitmap_path_32 = String::formatted("/res/icons/hackstudio/templates-32x32/{}.png", id);
+    auto bitmap_path_16 = String::formatted("/res/icons/hackstudio/templates-16x16/{}.png", id);
+
+    if (Core::File::exists(bitmap_path_32)) {
+        auto bitmap32 = Gfx::Bitmap::load_from_file(bitmap_path_32);
+        auto bitmap16 = Core::File::exists(bitmap_path_16)
+            ? Gfx::Bitmap::load_from_file(bitmap_path_16)
+            : Gfx::Bitmap::load_from_file("/res/icons/16x16/filetype-executable.png");
+
+        icon = GUI::Icon(move(bitmap16), move(bitmap32));
+    }
```

Comment:
- Could we make this explicit and add the (32px) icon to the config file as well?
- Do we really need to attempt loading the 16px icon, that doesn't actually exist and is never shown in the widget anyway, as AFAICT?

Reply:
Scrapping the 16x loading is definitely a good idea. I included it initially in the interest of completeness, but in hindsight simpler would definitely be better here.